pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+    }
+
+  Both these ideas are questionable in the context of this proposal. See
+  the section on `rejected`_ ideas.
+
+
+.. _specification:
+
+Specification
+=============
+
+Terminology
+-----------
+
+We propose to use the term *protocols* for types supporting structural
+subtyping. The reason is that the term *iterator protocol*,
+for example, is widely understood in the community, and coming up with
+a new term for this concept in a statically typed context would just create
+confusion.
+
+This has the drawback that the term *protocol* becomes overloaded with
+two subtly different meanings: the first is the traditional, well-known but
+slightly fuzzy concept of protocols such as iterator; the second is the more
+explicitly defined concept of protocols in statically typed code.
+The distinction is not important most of the time, and in other
+cases we propose to just add a qualifier such as *protocol classes*
+when referring to the static type concept.
+
+If a class includes a protocol in its MRO, the class is called
+an *explicit* subclass of the protocol. If a class is a structural subtype
+of a protocol, it is said to implement the protocol and to be compatible
+with a protocol. If a class is compatible with a protocol but the protocol
+is not included in the MRO, the class is an *implicit* subtype
+of the protocol.
+
+The attributes (variables and methods) of a protocol that are mandatory
+for other class in order to be considered a structural subtype are called
+protocol members. Other members defined in a protocol class are called
+non-protocol members.
+
+
+.. _definition:
+
+Defining a protocol
+-------------------
+
+Protocols are defined by including a special new class ``typing.Protocol``
+(an instance of ``abc.ABCMeta``) in the base classes list, preferably
+at the beginning of the list. Here is a simple example::
+
+  from typing import Protocol
+
+  class SupportsClose(Protocol):
+      @abstractmethod
+      def close(self) -> None:
+          ...
+
+Now if one defines a class ``Resource`` with a ``close()`` method that has
+a compatible signature, it would implicitly be a subtype of
+``SupportsClose``, since the structural subtyping is used for
+protocol types::
+
+  class Resource:
+      ...
+      def close(self) -> None:
+          self.file.close()
+          self.lock.release()
+
+Apart from few restrictions explicitly mentioned below, protocol types can
+be used in every context where a normal types can::
+
+  def close_all(things: Iterable[SupportsClose]) -> None:
+      for t in things:
+          t.close()
+
+  f = open('foo.txt')
+  r = Resource()
+  close_all([f, r])  # OK!
+  close_all([1])     # Error: 'int' has no 'close' method
+
+Note that both the user-defined class ``Resource`` and the built-in
+``IO`` type (the return type of ``open()``) are considered subtypes of
+``SupportsClose``, because they provide a ``close()`` method with
+a compatible type signature.
+
+
+Protocol members
```

Comment:
> This seems to imply that sequences no longer support iteration, since Sequence.__iter__ is not abstract and thus is not a protocol member. Is this intentional?

``Sequence`` _implements_ ``__iter__`` therefore it will be accepted wherever ``Iterable`` is expected. But asking for ``__iter__`` on ``Sequence`` indeed will fail.

This indeed looks problematic. Here is a solution that I could propose: we do not have such thing as non-protocol members. However, there will be two kinds of members, abstract and non-abstract. _All_ of them should be implemented in order to consider a class as structural (implicit) subtype.

The reason to have non-abstract methods is that one can get them "for free" using explicit subclassing provided abstract methods are implemented in subclass. So that for larger protocols like ``Sequence`` or ``Mapping`` one doesn't need to do
```python
def method(self):
    return super().method()
```
for every method.

This coincides with how ABCs work at runtime, so that this will be intuitive for someone who already worked with ABCs.